sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -1425,11 +1436,33 @@ class Analyzer(
       // rule: ResolveDeserializer.
       case plan if containsDeserializer(plan.expressions) => plan
 
-      // SPARK-25942: Resolves aggregate expressions with `AppendColumns`'s children, instead of
-      // `AppendColumns`, because `AppendColumns`'s serializer might produce conflict attribute
-      // names leading to ambiguous references exception.
-      case a @ Aggregate(groupingExprs, aggExprs, appendColumns: AppendColumns) =>
-        a.mapExpressions(resolveExpressionTopDown(_, appendColumns))
+      case a: Aggregate =>
+        val planForResolve = a.child match {
+          case appendColumns: AppendColumns => appendColumns
+          case _ => a
+        }
+
+        val resolvedGroupingExprs = a.groupingExpressions
+          .map(resolveExpressionTopDown(_, planForResolve, true, false))
+
+        val resolvedAggExprs = a.aggregateExpressions
+          .map(resolveExpressionTopDown(_, planForResolve, true, true))
+            .map(_.asInstanceOf[NamedExpression])
```

Comment:
I'd expect something simpler
```
// If `trimAlias` is true, trim Aliases over non-top-level GetStructField.
def resolveExpressionTopDown(..., trimAlias: Boolean = false, isTopLevel: Boolean = true) = {
  ...
  result match {
    case Alias(s: GetStructField, _) if trimAlias && !isTopLevel => s
    case _ => result
  }
  ...
  case _ => e.mapChildren(resolveExpressionTopDown(_, q, trimAlias, isTopLevel = false))
}
```

and here
```
val resolvedGroupingExprs = a.groupingExpressions
  .map(resolveExpressionTopDown(_, planForResolve, trimAlias = true))
  .map {
    // trim Alias over top-level GetStructField
    case Alias(s: GetStructField, _) => s
    case other => other
  }

val resolvedAggExprs = a.aggregateExpressions
  .map(resolveExpressionTopDown(_, planForResolve, trimAlias = true))
```

Reply:
Updated